Lib/asyncio/streams.py

```diff
@@ -81,16 +124,185 @@
         loop = events.get_event_loop()
 
     def factory():
-        reader = StreamReader(limit=limit, loop=loop,
-                              _asyncio_internal=True)
-        protocol = StreamReaderProtocol(reader, client_connected_cb,
-                                        loop=loop,
-                                        _asyncio_internal=True)
+        protocol = _LegacyServerStreamProtocol(limit,
+                                               client_connected_cb,
+                                               loop=loop,
+                                               _asyncio_internal=True)
         return protocol
 
     return await loop.create_server(factory, host, port, **kwds)
 
 
+class _BaseStreamServer:
+    # TODO: add __repr__
+
+    # Design notes.
+    # StreamServer and UnixStreamServer are exposed as FINAL classes, not function
+    # factories.
+    # async with serve(host, port) as server:
+    #      server.start_serving()
+    # looks ugly.
+    # The class doesn't provide API for enumerating connected streams
+    # It can be a subject for improvements in Python 3.9
+
+    def __init__(self, client_connected_cb,
+                 limit=_DEFAULT_LIMIT,
+                 shutdown_timeout=60,
+                 _asyncio_internal=False):
+        if not _asyncio_internal:
+            raise RuntimeError("_ServerStream is a private asyncio class")
+        self._client_connected_cb = client_connected_cb
+        self._limit = limit
+        self._loop = events.get_running_loop()
+        self._low_server = None
+        self._streams = {}
+        self._shutdown_timeout = shutdown_timeout
+
+    async def bind(self):
+        if self._low_server is not None:
+            return
+        self._low_server = await self._bind()
+
+    def is_bound(self):
+        return self._low_server is not None
+
+    def addresses(self):
+        # I don't want to expose plain socket.socket objects as low-level
+        # asyncio.Server does but exposing served IP addresses or unix paths
+        # is useful
+        #
+        # multiple value for socket bound to both IPv4 and IPv6 families
+        if self._low_server is None:
+            return []
+        return [sock.getsockname() for sock in self._low_server.sockets]
```

Comment:
I don't like the idea of returning lists, they cannot be cached (as users can mutate them). Let's return a tuple or a frozenset.

Also, in uvloop I return a "socket-like" object.  I think it makes more sense than adding multiple ad-hoc methods to introspect the underlying sockets.  Maybe we copy https://github.com/MagicStack/uvloop/blob/master/uvloop/pseudosock.pyx into asyncio 3.8?

Reply:
I completely understand why do you use PseudoSocket for uvloop implementation.
But for high-level API the only safe server socket operation is `sock.getsockname()` (and `sock.getsockopt()`. IPv4 vs IPv6 can be distinguished by the length of addresses item. Not very cool but works at least.

`setsockopt()` can have dangerous side-effects.
Even `fileno()` can be used for dirty hacks.
For high-level AIP, if we need the functionality provided by `setsockopt` better to support it by keyword-only parameters for stream constructor, as we already do for `backlog`, `reuse_address` and `reuse_port`.

I agree with `tuple` though. More restrictive type is better.
I don't like `frozenset` mostly because I would write `server.addresses()[0]`.

Reply:
> setsockopt() can have dangerous side-effects.
> Even fileno() can be used for dirty hacks.

Sometimes such hacks are necessary (e.g. people used it to set NODELAY before we had it).  I don't want to close the door here.  That's why I'm suggesting to return a socket-like object that allows *some* limited functionality, instead of adding `StreamServer.addresses`, `StreamServer.filenos`, `StreamServer.sockopts` etc -- because asyncio users will start demanding us to add them.

Reply:
`asyncio.Server.sockets` are returned.
The type is `list`, I didn't change the type of already existing `sockets` property